CONTRACTS: Support bounded user defined memory-predicates

remi-delmas-3000 opened the pull request:
Fixes https://github.com/diffblue/cbmc/issues/7194.

Users can now write their own pointer predicates in terms of:
- __CPROVER_is_fresh(ptr, size)
- __CPROVER_pointer_in_range(lb, ptr, ub)
- __CPROVER_obeys_contract(ptr, target)

A rewriting pass lifts these user-defined pointer predicates
into predicates that take pointer by reference and can
make the predicates hold in assumption contexts
using side-effects, and check the predicates in assertion
contexts.

Limitiations:
User-defined predicates can be self-recursive but not mutually
recursive, and their evaluation needs to be terminate, i.e. they
can only describe bounded data structures.

The doxygen user and developer manuals and has been updated to describe the feature and its implementation.
Tests were added, and existing tests cleaned-up.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,116 @@
+/// Collects all user-defined predicates that call functions
+/// is_fresh, pointer_in_range, obeys_contract and lifts them to function taking
+/// pointers by reference. When called in assumption contexts,
+/// These user-defined predicates update the pointers using side effect in order
```
```suggestion
/// these user-defined predicates update the pointers using side effect in order
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,116 @@
+  /// \param parameter_rank the parameter to update
+  /// \param replace_lifted_param symbol replacer (receives `p ~> *p` mappings)
+  /// The parameter symbol gets updated in the symbol table,
+  /// The function signature gets in the function gets updated with
+  /// the new type.
```
```suggestion
  /// The function signature gets updated with the new type.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,116 @@
+    message_handlert &message_handler);
+
+  /// \brief Collects and lifts all user-defined memory predicates.
+  /// \param[out] lifted set of predicates that were lifted
```
```suggestion
  /// \param[out] lifted Set of predicates that were lifted
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,116 @@
+
+  /// \brief Collects and lifts all user-defined memory predicates.
+  /// \param[out] lifted set of predicates that were lifted
+  /// \param[out] discovered_function_pointer_contracts set of function pointer
```
```suggestion
  /// \param[out] discovered_function_pointer_contracts Set of function pointer
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,116 @@
+    std::set<irep_idt> &discovered_function_pointer_contracts);
+
+  /// \brief adds a pointer_type to the parameter of a function
+  /// \param function_id the function
+  /// \param parameter_rank the parameter to update
+  /// \param replace_lifted_param symbol replacer (receives `p ~> *p` mappings)
```
```suggestion
  /// \param function_id The function
  /// \param parameter_rank The parameter to update
  /// \param replace_lifted_param Symbol replacer (receives `p ~> *p` mappings)
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-pointer-in-range.md:
```diff
@@ -0,0 +1,26 @@
+
+In goto programs encoding pre or post conditions (generated from the contract
+clauses) and in all user-defined functions, we simply replace calls to
+`__CPROVER_is_in_range` with calls to the library implementation
```
```suggestion
`__CPROVER_is_in_range` with calls to the library implementation.
```

remi-delmas-3000 replied:
fixed

remi-delmas-3000 replied:
fixed

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2342,6 +2342,26 @@ exprt c_typecheck_baset::do_special_functions(
+    if(expr.arguments().size() != 3)
+    {
+      throw invalid_source_file_exceptiont{
+        "is_in_range expects three arguments", expr.source_location()};
```
Could we cover these coverage holes?

remi-delmas-3000 replied:
added tests.

feliperodri commented on regression/contracts-dfcc/assigns-replace-malloc-zero/main.c:
```diff
@@ -1,37 +1,42 @@
 #include <stdlib.h>
+#include <stdbool.h>
+#include <assert.h>
```
We don't need this include.

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-fresh-requires-max-malloc-size/test.desc:
```diff
@@ -0,0 +1,12 @@
+This tests shows how assuming that a pointer is fresh with a nondeterministic
+size implicitly assumes that the size is strictly less than the max malloc
+size. This is in order to ensure that the internal representation used for
+assignable address ranges cannot suffer pointer overflows
```
newline

feliperodri replied:
```suggestion
assignable address ranges cannot suffer pointer overflows.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-in-range-requires/main.c:
```diff
@@ -0,0 +1,24 @@
+  size_t size;
+  char *cur;
+  foo(arr, size, cur);
+}
```
newline

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-in-range-requires/main.c:
```diff
@@ -0,0 +1,24 @@
+#include <stdbool.h>
```
We don't need this include.

feliperodri replied:
Could be replaced by <assert.h>

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-in-range-requires/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+--
+Checks that assuming pointer-in-range as preconditions establishes a state
+in which the definition of the predicate holds.
```
newline

remi-delmas-3000 replied:
fixed

tautschnig replied:
Not actually fixed?

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-in-range-several-targets/main.c:
```diff
@@ -0,0 +1,42 @@
+__CPROVER_ensures(__CPROVER_return_value == 0 || __CPROVER_return_value == 1)
+// clang-format on
+{
+CALL:
```
Why this?

remi-delmas-3000 replied:
to be able to locate the function pointer call by label when using `--restrict-function-pointer` in `test.desc`

feliperodri commented on regression/contracts-dfcc/memory-predicates-is-in-range-several-targets/main.c:
```diff
@@ -0,0 +1,42 @@
+#include <stdbool.h>
```
Unnecessary include.

remi-delmas-3000 replied:
fixed

tautschnig replied:
Still there?

feliperodri commented on regression/contracts-dfcc/memory-predicates-user-defined-requires-enforce/main.c:
```diff
@@ -0,0 +1,77 @@
+#include <assert.h>
```
Unnecessary include.

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-user-defined-requires-enforce/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+Checks if user-defined predicates work;
```
newline

feliperodri replied:
```suggestion
Checks if user-defined predicates work.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts-dfcc/memory-predicates-user-defined-requires-replace/main.c:
```diff
@@ -0,0 +1,95 @@
+#include <assert.h>
```
Unnecessary include.

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_in_range.h:
```diff
@@ -0,0 +1,56 @@
+
+/// \file
+/// Instruments occurrences of is_in_range predicates in programs
+/// encoding requires and ensures clauses of contracts
```
```suggestion
/// encoding requires and ensures clauses of contracts.
```

remi-delmas-3000 replied:
fixed

kroening commented:
A reader might believe that `__CPROVER_is_in_range` ought to apply to types other than pointers, e.g., integers, or floating-point numbers. It may make sense to make that predicate polymorphic. Can you simply use `__CPROVER_pointer_in_range`?

feliperodri commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
````diff
@@ -0,0 +1,209 @@
+__CPROVER_bool __CPROVER_obeys_contract(void (*fptr)(void), void (*contract)(void));
+```
+
+Users are free to express pre and post conditions using these predicates,
+and we let users define their own functions in terms of these predicates.
````
Important to make clear here if users could use them in general assertions/assumptions.

remi-delmas-3000 replied:
added a sentence in both user and developer doc.

feliperodri commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,209 @@
+
+The last step of the rewriting is to apply the normal DFCC instrumentation
+which adds a write set parameter to the function, instruments it for side
+effect cheching, and maps core memory predicates to their implementations.
```
```suggestion
effect checking, and maps core memory predicates to their implementations.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,209 @@
+The write_set parameter carries assumption/assertion context flags, so that the
+core predicate implementation functions for `__CPROVER_is_fresh`,
+`__CPROVER_is_in_range` and `__CPROVER_obeys_contract` know when to
+update the pointers in place using malloc and assignents to make the predicates
```
```suggestion
update the pointers in place using `malloc` and assignments to make the predicates
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/doc/user/contracts-memory-predicates.md:
````diff
@@ -119,6 +120,117 @@ int foo()
+bool __CPROVER_is_in_range(void *lb, void *p, void *ub);
+```
+
+This predicate holds if `lb`, `p` and `ub` are valid pointer within the same
````
```suggestion
This predicate holds if `lb`, `p` and `ub` are valid pointers within the same
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/doc/user/contracts-memory-predicates.md:
```diff
@@ -119,6 +120,117 @@ int foo()
+
+### Limitations
+
+The main limitation with user defined predicates is that they must be bounded.
+In the `is_list` example above, recursion is bounded by the use of the `len`
+parameter.
+The `is_double_buffer` predicate also describes a bounded structure.
```
Do we report any errors in case the predicate is not bounded or CBMC will just try to unroll it forever.

remi-delmas-3000 replied:
For now it is in the user's hands to ensure boundedness. Symex will just run forever and the user will have to ctrl-c it.

remi-delmas-3000 replied:
We could later impose syntactic restrictions like guarded recursion and implement a lazy expansion of the predicates (automatically upon access or using open/close statements like in verifast or similar tools).

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.cpp:
```diff
@@ -0,0 +1,425 @@
+
+#include "dfcc_lift_memory_predicates.h"
+
+#include <util/cprover_prefix.h>
+#include <util/format_expr.h>
+#include <util/graph.h>
+#include <util/pointer_expr.h>
+#include <util/replace_symbol.h>
+#include <util/symbol.h>
+
+#include <goto-programs/goto_model.h>
+
```
Is this order right?

remi-delmas-3000 replied:
yes, that's the order produced by the clang-format rules.

feliperodri reviewed (changes requested):
Only some minor changes before approval.

remi-delmas-3000 commented:
> A reader might believe that `__CPROVER_is_in_range` ought to apply to types other than pointers, e.g., integers, or floating-point numbers. It may make sense to make that predicate polymorphic. Can you simply use `__CPROVER_pointer_in_range`?

Will fix.

EDIT: 
My instrumentation pass maps predicate calls to built-in predicates to special implementations that takes the dynamic frame as parameter and to make it context-aware (assert vs. assume).

Right now the front-end maps `__CPROVER_pointer_in_range` calls to `pointer_in_range_exprt` expressions that get lowered during goto-conversion, so when my pass happens `pointer_in_range_exprt` is not there for me to match and replace anymore.


@tautschnig is currently working on moving the lowering of built-ins to the back end #7395, which would solve my problem by keeping `pointer_in_range_exprt` intact out of goto-conversion.


The best I can do for the moment is to temporarily rename my `__CPROVER_is_in_range` predicate to `__CPROVER_pointer_in_range_dfcc` so it does not get lowered during goto conversion and my pass can still work.

I'll update my code to `pointer_in_range_exprt` when #7395 gets merged.


@kroening  Would that work for you ?

tautschnig commented on regression/contracts-dfcc/memory-predicates-is-fresh-requires-max-malloc-size/test.desc:
```diff
@@ -0,0 +1,12 @@
+This tests shows how assuming that a pointer is fresh with a nondeterministic
+size implicitly assumes that the size is strictly less than the max malloc
+size. This is in order to ensure that the internal representation used for
+assignable address ranges cannot suffer pointer overflows.
```
Newline

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2334,6 +2334,26 @@ exprt c_typecheck_baset::do_special_functions(
+      }
+    }
+    return nil_exprt();
+  }
```
Can this (and any other support for `__CPROVER_is_in_range`) please go in a commit of its own? #7395 should make it completely unnecessary.

tautschnig replied:
I think this would be in line with what's suggested in https://github.com/diffblue/cbmc/pull/7401#issuecomment-1341612884 (except it would make the follow-up work very easy to identify).

remi-delmas-3000 replied:
I moved all pointer_in_range related code to a separate commit

tautschnig commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,210 @@
+1. in assumption contexts, evaluating the predicate allocates the data stucture
+  specified by the predicate;
+1. in assertion contexts, evaluating the predicate checks that a given pointer
+  satisfies the predicate definiton;
```
s/definiton/definition/

remi-delmas-3000 replied:
FIXED

tautschnig commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+By rewriting such user-defined predicate we achieve two things:
+1. in assumption contexts, evaluating the predicate allocates the data stucture
+  specified by the predicate;
````
What allocation depth does this imply?

remi-delmas-3000 replied:
The allocation depth has to be bounded by the predicate definition itself. i.e. for a list the predicate should have a user-defined depth parameter.

tautschnig commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,210 @@
+By rewriting such user-defined predicate we achieve two things:
+1. in assumption contexts, evaluating the predicate allocates the data stucture
+  specified by the predicate;
+1. in assertion contexts, evaluating the predicate checks that a given pointer
```
Should this be "2.?"

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,210 @@
+  satisfies the predicate definiton;
+
+To be able to achieve point 1., the user-defined functions need to be
+transformed into functions that take their pointer arguments by reference
```
"need to be transformed" -- please avoid passive voice so as to make clear who actually does this.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/doc/developer/contracts-dev-spec-memory-predicates-rewriting.md:
```diff
@@ -0,0 +1,210 @@
+
+The set of predicates is rewritten in topological call-dependency order, so that
+when instrumenting a predicate P, any other predicate Q it calls is already
+instrumented and we know which of Q parameters have been lifted.
```
Recursion?

remi-delmas-3000 replied:
Added explanation on how self-recursion and mutual recursion are detected.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc.cpp:
```diff
@@ -269,6 +273,18 @@ void dfcct::instrument_harness_function()
+  memory_predicates.lift_predicates(predicates, function_pointer_contracts);
+  for(const auto &predicate : predicates)
+  {
+    log.debug() << "COUCOU Memory predicate" << predicate << messaget::eom;
```
What does `COUCOU` stand for? Also, wouldn't you want a space before the actual predicate?

remi-delmas-3000 replied:
Coucou means Hello in French. Debugging leftover.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.cpp:
```diff
@@ -0,0 +1,425 @@
+}
+
+void dfcc_lift_memory_predicatest::lift_predicates(
+  std::set<irep_idt> &lifted,
```
I suggest you make this the return value. There is only one call site, and it seemingly will work just fine when getting back a set, no need to incrementally update an existing set.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.cpp:
```diff
@@ -0,0 +1,425 @@
+  const symbolt &old_symbol = entry->second;
+  const auto &old_symbol_expr = old_symbol.symbol_expr();
+  // create new parameter symbol, same everything except type
+  symbolt new_symbol;
```
Please try to use one of the new `symbolt` constructors so that at least the name, type, and mode are set at construction time.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.cpp:
```diff
@@ -0,0 +1,425 @@
+      }
+    }
+  }
+  for(auto instruction : body.instructions)
+  {
+    instruction.output(log.debug());
+  }
```
Please use `conditional_output`

remi-delmas-3000 replied:
fixed in the whole file

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,115 @@
+  dfcc_utilst &utils;
+  dfcc_libraryt &library;
+  dfcc_instrumentt &instrument;
+  message_handlert &message_handler;
+  messaget log;
```
Do you actually need both of these? You could always get the message handler from `log` using `get_message_handler()`.

remi-delmas-3000 replied:
Fixed

Commit:
CONTRACTS: support bounded user-defined memory predicates

Users can now write their own pointer predicates in terms of:
- __CPROVER_is_fresh(ptr, size)
- __CPROVER_obeys_contract(ptr, target)

A rewriting pass lifts these user-defined pointer predicates
into predicates that take pointers by reference and can
make the predicates hold in assumption contexts using
side-effects, and check the predicates in assertion contexts.

Limitiations:
User-defined predicates can be self-recursive but not mutually
recursive, and their evaluation needs to be terminate, i.e. they
can only describe bounded data structures.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_lift_memory_predicates.h:
```diff
@@ -0,0 +1,113 @@
+  /// signature gets updated with the new type.
+  void add_pointer_type(
+    const irep_idt &function_id,
+    const std::size_t &parameter_rank,
```
Nit pick: I wouldn't pass a POD by reference, unless you have some very good reasons.

remi-delmas-3000 replied:
fixed

Commit:
CONTRACTS: support pointer_in_range_dfcc predicate

Adds support for pointer_in_range_dfcc in ensures and
requires clauses and user-defined predicates. This is a
temporary workaround to the fact that pointer_in_range
is lowered by the front-end.

feliperodri commented on regression/contracts-dfcc/memory-predicates-pointer-in-range-requires/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+--
+Checks that assuming pointer-in-range as preconditions establishes a state
+in which the definition of the predicate holds.
```
newline

remi-delmas-3000 replied:
fixed

remi-delmas-3000 commented:
@feliperodri all requested changes were applied

feliperodri reviewed (approved):
LGTM
